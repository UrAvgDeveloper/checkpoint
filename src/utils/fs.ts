import * as fs from 'fs';
import * as util from 'util';

export const exists = util.promisify(fs.exists);
export const mkdir = util.promisify(fs.mkdir);
