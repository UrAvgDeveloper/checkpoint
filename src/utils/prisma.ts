import { GraphQLObjectType, GraphQLField } from 'graphql';

export const transformEntityObjectsToPrismaSchema = async (
  entities: GraphQLObjectType[]
): Promise<string> => {
  const models: string[] = [];
  for (const entity of entities) {
    let model = `model ${entity.name} {\n`;

    Object.values(entity.getFields()).forEach(field => {
      model += `${transformEntityFieldToPrismaSchema(field)}\n`;
    });
    model += '}';
    models.push(model);
  }

  return models.join('\n\n');
};

const transformEntityFieldToPrismaSchema = async (
  field: GraphQLField<any, any>
): Promise<string> => {
  // algorithms:
  //
  // map unwrapped graphql type to prisma type
  //
  // if type is a list and scalar type
  // make the type a list type
  //
  // else include a relationships directive in the prisma schema
  // the foreign key should be the id (think more about this).
  return '';
};
