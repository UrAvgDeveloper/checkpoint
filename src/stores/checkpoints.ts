import * as crypto from 'crypto';
import { Logger } from '../utils/logger';

type ToString = {
  toString: () => string;
};

export interface CheckpointRecord {
  blockNumber: number;
  contractAddress: string;
}

/**
 * Metadata Ids stored in the CheckpointStore.
 *
 */
export enum MetadataId {
  LastIndexedBlock = 'last_indexed_block'
}

const CheckpointIdSize = 10;

/**
 * Generates a unique hex based on the contract address and block number.
 * Used when as id for storing checkpoints records.
 *
 */
export const getCheckpointId = (contract: string, block: number): string => {
  const data = `${contract}${block}`;
  return crypto.createHash('sha256').update(data).digest('hex').slice(-CheckpointIdSize);
};

/**
 * Checkpoints store is a data store class for managing
 * checkpoints data schema and records.
 *
 * It interacts with an underlying mysql database.
 */
export class CheckpointsStore {
  private readonly log: Logger;

  constructor(private readonly prisma: any, log: Logger) {
    this.log = log.child({ component: 'checkpoints_store' });
  }

  public async getMetadata(id: string): Promise<string | null> {
    const result = await this.prisma.metadata.findFirst({
      where: { id },
      select: {
        value: true
      }
    });

    if (!result || result.value.length == 0) {
      return null;
    }

    return result.value[0].value;
  }

  public async getMetadataNumber(id: string, base = 10): Promise<number | undefined> {
    const strValue = await this.getMetadata(id);
    if (!strValue) {
      return undefined;
    }

    return parseInt(strValue, base);
  }

  public async setMetadata(id: string, value: ToString): Promise<void> {
    await this.prisma.metadata.upsert({
      where: { id },
      update: {
        value: value.toString()
      },
      create: {
        id,
        value: value.toString()
      }
    });
  }

  public async insertCheckpoints(checkpoints: CheckpointRecord[]): Promise<void> {
    if (checkpoints.length === 0) {
      return;
    }

    await this.prisma.checkpoint.createMany({
      data: checkpoints.map(checkpoint => {
        const id = getCheckpointId(checkpoint.contractAddress, checkpoint.blockNumber);
        return {
          id,
          block_number: checkpoint.blockNumber,
          contract_address: checkpoint.contractAddress
        };
      }),
      skipDuplicates: true
    });
  }

  /**
   * Fetch list of checkpoint blocks greater than or equal to the
   * block number arguments, that have some events related to the
   * contracts in the lists.
   *
   * By default this returns at most 15 next blocks. This return limit
   * can be modified by the limit command.
   */
  public async getNextCheckpointBlocks(
    block: number,
    contracts: string[],
    limit = 15
  ): Promise<number[]> {
    const result = await this.prisma.checkpoint.findMany({
      where: {
        block_number: {
          gte: block
        },
        contract_address: {
          in: contracts
        }
      },
      orderBy: {
        block_number: 'asc'
      },
      take: limit
    });

    this.log.debug({ result, block, contracts }, 'next checkpoint blocks');

    return result.map(value => Number(value.block_number));
  }
}
