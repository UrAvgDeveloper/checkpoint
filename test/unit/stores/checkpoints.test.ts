import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { CheckpointsStore, getCheckpointId } from '../../../src/stores/checkpoints';
import { Logger } from '../../../src/utils/logger';

describe('CheckpointsStore', () => {
  let store: CheckpointsStore;
  let mockPrisma: any;
  let logger: Logger & DeepMockProxy<Logger>;

  beforeEach(() => {
    mockPrisma = mockDeep();
    logger = mockDeep<Logger>({
      child: () => mockDeep()
    });
    store = new CheckpointsStore(mockPrisma, logger);
  });

  describe('insertCheckpoints', () => {
    it('should should execute correct query', async () => {
      const checkpoints = [
        {
          contractAddress: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3',
          blockNumber: 5000
        },
        {
          contractAddress: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3',
          blockNumber: 123222
        }
      ];
      await store.insertCheckpoints(checkpoints);

      expect(mockPrisma.checkpoint.createMany.mock.calls).toMatchSnapshot();

      // verify id is properly computed
      const { id } = mockPrisma.checkpoint.createMany.mock.calls[0][0].data[0];

      expect(id).toEqual(
        getCheckpointId(checkpoints[0].contractAddress, checkpoints[0].blockNumber)
      );
    });
  });
});
