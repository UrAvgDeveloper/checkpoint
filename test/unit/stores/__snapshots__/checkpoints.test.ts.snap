// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckpointsStore insertCheckpoints should should execute correct query 1`] = `
Array [
  Array [
    Object {
      "data": Array [
        Object {
          "block_number": 5000,
          "contract_address": "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
          "id": "a739beda26",
        },
        Object {
          "block_number": 123222,
          "contract_address": "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
          "id": "d2b8dcc2b5",
        },
      ],
      "skipDuplicates": true,
    },
  ],
]
`;
